### PR TITLE
Warn when taking address of local or parameter in async method

### LINKED ISSUE #66829: Compiler doesnt create field in StateMachine in release mode
**Version Used**: dotnet 6.0.308

**Steps to Reproduce**:

1. Write code below
2. Run in debug
3. Run in release

It happens if pointers is used

```csharp
long x = 10;

unsafe
{
    Console.WriteLine(*&x);
}

await Task.Delay(1000);

unsafe
{
    Console.WriteLine(*&x);
}
```

**Expected Behavior**: prints 10 both times

**Actual Behavior**: prints 10 and 0

### PR DESCRIPTION
**Pull request by RikkiGibson**
Closes #66829 and #63100.

Using `&` on a local or parameter which may be hoisted to a class field is a GC hole. Rather than design a solution where such variables can be used in `fixed` statements, or are automatically `fixed`, we introduce a warning for taking the address of 
these variables.

We judge it to be exceedingly rare for users to want to do this, and other state-machine-backed constructs like iterators completely disallow unsafe code. If we were starting from scratch here, we would probably also have errors for unsafe code in async methods. We introduce the warning in order to maintain compat and to guide users away from using the problematic coding pattern moving forward.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/IteratorAndAsyncCaptureWalker.cs [hunk 1]
```diff
@@ -317,6 +317,13 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
             return null;
         }
 
+        public override BoundNode VisitAddressOfOperator(BoundAddressOfOperator node)
+        {
+            CheckAssigned(node.Operand, node.Operand.Syntax);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/IteratorAndAsyncCaptureWalker.cs:322]: See doc comment on containing type for more context.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs [hunk 2]
```diff
@@ -871,6 +871,97 @@ public static void Main()
...
+
+                class Program
+                {
+                    public static async Task Main()
+                    {
+                        long x = 1;
+
+                        unsafe
+                        {
+                            Console.Write(*&x);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs:889]: Note that unsafe is not allowed at all in iterators so no scenario to test there.

**Comment by RikkiGibson** [comment]
Would like to pause on this change until we know what to do about taking the address of variables which are actually fields.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs [hunk 3]
```diff
@@ -871,6 +871,167 @@ public static void Main()
...
+                        }
+
+                        unsafe
+                        {
+                            Console.Write(*&x);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs:894]: We should also verify that a non-async local function inside an async method does not warn when `&` is used.

**Comment by RikkiGibson** [comment]
@dotnet/roslyn-compiler for review

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs [hunk 4]
```diff
@@ -871,6 +871,204 @@ public static void Main()
...
+                //             Console.Write(*&x);
+                Diagnostic(ErrorCode.WRN_AddressOfInAsync, "x").WithLocation(12, 29),
+                // (19,29): warning CS9123: The '&' operator should not be used on parameters or local variables in async methods.
+                //             Console.Write(*&x);
+                Diagnostic(ErrorCode.WRN_AddressOfInAsync, "x").WithLocation(19, 29)
+            };
+
+            CompileAndVerify(source, options: TestOptions.UnsafeDebugExe.WithMetadataImportOptions(MetadataImportOptions.All), expectedOutput: "11", symbolValidator: debugSymbolValidator, verify: Verification.Fails)
+                .VerifyDiagnostics(diagnostics);
+            CompileAndVerify(source, options: TestOptions.UnsafeReleaseExe.WithMetadataImportOptions(MetadataImportOptions.All), expectedOutput: "10", symbolValidator: releaseSymbolValidator, verify: Verification.Fails)
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs:972]: >0 [](http://example.com/codeflow?start=147&length=1)

Why is there a difference between debug and release?

Comment (cston) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs:972]: Actually, it looks like this is [https://github.com/dotnet/roslyn/issues/66829](https://github.com/dotnet/roslyn/issues/66829).

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs [hunk 5]
```diff
@@ -871,6 +871,204 @@ public static void Main()
...
+
+            void releaseSymbolValidator(ModuleSymbol module)
+            {
+                var stateMachine = module.GlobalNamespace.GetMember<NamedTypeSymbol>("Program.<Main>d__0");
+                // Test that there is no state-machine field based on 'x'.
+                Assert.Empty(stateMachine.GetMembers().Where(m => m.Name.StartsWith("<x>")));
+            }
+        }
+
+        [Fact]
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs:932]: >Fact [](http://example.com/codeflow?start=9&length=4)

Consider adding `[WorkItem(66829)]`.

**Review by cston — APPROVED** [review]

**Comment by RikkiGibson** [comment]
@dotnet/roslyn-compiler for a second review of a fairly small implementation change.

**Review by jjonescz — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs [hunk 6]
```diff
@@ -206,6 +206,10 @@ internal static int GetWarningLevel(ErrorCode code)
             // docs/compilers/CSharp/Warnversion Warning Waves.md
             switch (code)
             {
+                case ErrorCode.WRN_AddressOfInAsync:
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs:209]: Was the warning wave document updated correspondingly? 
https://github.com/dotnet/roslyn/blob/main/docs/compilers/CSharp/Warnversion%20Warning%20Waves.md

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs:209]: Opened #67420

**Pull request merged**